ArduCopter/takeoff.cpp

```diff
@@ -48,95 +48,50 @@ bool Mode::do_user_takeoff(float takeoff_alt_cm, bool must_navigate)
     return true;
 }
 
-// start takeoff to specified altitude above home in centimeters
+// start takeoff to specified altitude above home in centimetres
 void Mode::_TakeOff::start(float alt_cm)
 {
     // indicate we are taking off
     copter.set_land_complete(false);
     // tell position controller to reset alt target and reset I terms
     copter.flightmode->set_throttle_takeoff();
 
-    // calculate climb rate
-    const float speed = MIN(copter.wp_nav->get_default_speed_up(), MAX(copter.g.pilot_speed_up*2.0f/3.0f, copter.g.pilot_speed_up-50.0f));
-
-    // sanity check speed and target
-    if (speed <= 0.0f || alt_cm <= 0.0f) {
-        return;
-    }
-
     // initialise takeoff state
     _running = true;
-    max_speed = speed;
-    start_ms = millis();
-    alt_delta = alt_cm;
+    take_off_start_alt = copter.pos_control->get_pos_target_z_cm();
+    take_off_alt = take_off_start_alt + alt_cm;
 }
 
 // stop takeoff
 void Mode::_TakeOff::stop()
 {
     _running = false;
-    start_ms = 0;
 }
 
-// returns pilot and takeoff climb rates
-//  pilot_climb_rate is both an input and an output
-//  takeoff_climb_rate is only an output
-//  has side-effect of turning takeoff off when timeout as expired
-void Mode::_TakeOff::get_climb_rates(float& pilot_climb_rate, float& takeoff_climb_rate)
+// do_pilot_takeoff - commands the aircraft to the take off altitude
+//  climb is cancelled if pilot_climb_rate_cm becomes negative
+//  sets take off to complete when target altitude is within 1% of the take off altitude
+void Mode::_TakeOff::do_pilot_takeoff(float& pilot_climb_rate_cm)
 {
     // return pilot_climb_rate if take-off inactive
     if (!_running) {
-        takeoff_climb_rate = 0.0f;
         return;
     }
 
-    // acceleration of 50cm/s/s
-    static constexpr float TAKEOFF_ACCEL = 50.0f;
-    const float takeoff_minspeed = MIN(50.0f, max_speed);
-    const float time_elapsed = (millis() - start_ms) * 1.0e-3f;
-    const float speed = MIN(time_elapsed * TAKEOFF_ACCEL + takeoff_minspeed, max_speed);
+    Vector3f pos;
+    Vector3f vel;
+    Vector3f accel;
 
-    const float time_to_max_speed = (max_speed - takeoff_minspeed) / TAKEOFF_ACCEL;
-    float height_gained;
-    if (time_elapsed <= time_to_max_speed) {
-        height_gained = 0.5f * TAKEOFF_ACCEL * sq(time_elapsed) + takeoff_minspeed * time_elapsed;
-    } else {
-        height_gained = 0.5f * TAKEOFF_ACCEL * sq(time_to_max_speed) + takeoff_minspeed * time_to_max_speed +
-                        (time_elapsed - time_to_max_speed) * max_speed;
-    }
```

Comment:
Because people hack shit in without understanding what they are doing. This has been done all over the place and is one of the main reasons this rewrite was so difficult.

In all fairness what they did worked so I can respect that :)

Reply:
At the time this did change the existing behaviour however.